Fix: Rootless Podman-in-Podman on WSL

dvorst opened the pull request:
Fixes: #27411

## pkg/machine/wsl/declares.go
Adjust SUB_UID and SUB_GID ranges to support running rootless Podman inside a rootless run Podman container.

By default, a new user is assigned the following sub-ID ranges:
  SUB_UID_MIN=100000, SUB_GID_MIN=100000, SUB_UID_COUNT=65536, SUB_GID_COUNT=65536
This means the user’s sub-UID and sub-GID ranges are 100000–165535.

When the container is run rootless by the user in WSL, ID mappings occur as follows:
- Container ID 0 (root) maps to user ID 1000 on the host (user in WSL).
- Container IDs 1–65536 map to IDs 100000–165535 on host (the subid range previously mentioned).

If a new user is created inside this container (to build containers for example), it will attempt to use the default sub-ID range (100000–165535). However, this exceeds the container’s available ID mapping, since only IDs up to 65536 are mapped, causing rootless PinP to fail.

To enable container-in-container builds, the sub-ID ranges for the user in WSL must be large enough to provide at least 65536 usable IDs. A minimum SUB_UID_COUNT and SUB_GID_COUNT of 165536 is thus required, but 200000 is used here to provide additional margin.

<!--
Thanks for sending a pull request!

For more detailed information, please review our contributing guidelines:
https://github.com/containers/podman/blob/main/CONTRIBUTING.md#submitting-pull-requests
-->

#### Checklist

Ensure you have completed the following checklist for your pull request to be reviewed:
<!-- Use [x] to mark as done, or click the checkbox after opening PR -->

- [x] Certify you wrote the patch or otherwise have the right to pass it on as an open-source patch by signing all
commits. (`git commit -s`). (If needed, use `git commit -s --amend`).  The author email must match
the sign-off email address. See [CONTRIBUTING.md](https://github.com/containers/podman/blob/main/CONTRIBUTING.md#sign-your-prs)
for more information.
- [x] Referenced issues using `Fixes: #00000` in commit message (if applicable)
- [x] [Tests](https://github.com/containers/podman/tree/main/test#readme) have been added/updated (or no tests are needed)
- [x] [Documentation](https://github.com/containers/podman/blob/main/docs/README.md) has been updated (or no documentation changes are needed)
- [x] All commits pass `make validatepr` (format/lint checks)
- [x] [Release note](https://github.com/kubernetes/community/blob/master/contributors/guide/release-notes.md) entered in the section below (or `None` if no user-facing changes)

#### Does this PR introduce a user-facing change?

<!--
Write `None` if there are no user-facing changes, otherwise enter your release note below.
Include "action required" if users need to take action when upgrading.
-->

```release-note
Fixed Rootless Podman-in-Podman on WSL, by sufficiently increasing the SUB_GID and SUB_UID range of the user on WSL (#27411).
```

Luap99 reviewed (commented):
You need to rebase your PR to pass the search failure in the integration tests (not related to your code).

And please squash both commits into one, we prefer code and test in the same commits.

dvorst commented on pkg/machine/wsl/declares.go:
```diff
@@ -38,6 +38,8 @@ ln -fs /dev/null /etc/systemd/system/console-getty.service
 mkdir -p /etc/systemd/system/systemd-sysusers.service.d/
 echo CREATE_MAIL_SPOOL=no >> /etc/default/useradd
+sed -ir 's/SUB_UID_COUNT.*/SUB_UID_COUNT 200000/' /etc/login.defs
+sed -ir 's/SUB_GID_COUNT.*/SUB_GID_COUNT 200000/' /etc/login.defs
```
Thanks for the intermediate review.

I'll change the count to `1 000 000` instead of `200 000`. I’m quite busy with work at the moment, but I might have time to work on this around Christmas.

I'll also modify `/etc/subuid` and `/etc/subgid` directly, right after the adduser command is run, to align it with the `ignition.go` script.

My initial approach was based on the fact that adduser doesn’t provide an option to configure the sub-ID ranges, so these had to be set in a different way. Changing the defaults seemed to be a cleaner approach than modifying subuid/subgid directly, since that will give issues if adduser/useradd behavior ever changes (though i don't think that will ever change without backwards compatibility).

dvorst commented on pkg/machine/e2e/init_test.go:
```diff
@@ -124,6 +125,62 @@ var _ = Describe("podman machine init", func() {
+			n, err := strconv.Atoi(count[1])
+			Expect(err).ToNot(HaveOccurred())
+			Expect(n).To(BeNumerically(">=", count_min), "expected last number %d to be >= %d", n, count_min)
+		}
```
There is a count for the user and group id, so there will be a for loop anyway. The comment is meant to be informative for if a second user is ever created, and that this test will basically be run for all users.

dvorst committed:
Fix: Rootless Podman-in-Podman on WSL

closes: #27411

Adjust SUB_UID and SUB_GID ranges to support running rootless Podman inside a rootless run Podman container.
Also add a test to verify the change and prevent regression.

By default, a new user is assigned the following sub-ID ranges:
  SUB_UID_MIN=100000, SUB_GID_MIN=100000, SUB_UID_COUNT=65536, SUB_GID_COUNT=65536
This means the user’s sub-UID and sub-GID ranges are 100000–165535.

When the container is run rootless with the user defined below, ID mappings occur as follows:
- Container ID 0 (root) maps to user ID 1000 on the host (which is the user created below).
- Container IDs 1–65536 map to IDs 100000–165535 on host (the subid range previously mentioned).

If a new user is created inside this container (to build containers for example), it will
attempt to use the default sub-ID range (100000–165535). However, this exceeds the container’s
available ID mapping, since only IDs up to 65536 are mapped. This causes nested rootless Podman
to fail.

To enable container-in-container builds, the sub-ID ranges for the user must be large enough
to provide at least 65536 usable IDs. A minimum SUB_UID_COUNT and SUB_GID_COUNT of 165536 is
required, but 1,000,000 is used here to provide additional margin.

1,000,000 matches the subid range other machines are using, defined in [ignition.go](https://github.com/containers/podman/blob/69b397af49acd595b8d5b36971988d49951c043a/pkg/machine/ignition/ignition.go#L284-L289).
The script of other machines modify the subid files directly for 1 user, the `sed` command used in
this fix mimics that.

The test is added as en extension to the 'simple init with username' test case, to prevent having
to create a new VM.

Signed-off-by: dvorst <87502756+dvorst@users.noreply.github.com>

Luap99 reviewed (approved):
LGTM
